heat/core/io.py

```diff
@@ -354,36 +395,181 @@ def save_netcdf(data, path, variable, mode="w", **kwargs):
+                return ex_ind_nonempty[split_sq]
+            # split along empty dimension: split doesnt matter, only one process contains data
+            # return the last empty dimension (in expanded shape) before (the first nonempty dimension after split)
+            # number of nonempty elems before split
+            ne_before_split = split - shape[:split].count(1)
+            ind_ne_after_split = ind_nonempty[
+                ne_before_split
+            ]  # index of (first nonempty element after split) in squeezed shape
+            return max(
+                i
+                for i, v in enumerate(expandedShape[: max(ex_ind_nonempty[:ind_ne_after_split])])
+                if v == 1
+            )
+
+        def mergeSlices(var, var_slices, data, data_slices=None):
+            """ Using var[var_slices][data_slices] = data
+            combines a __getitem__ with a __setitem__ call, therefore it does not allow
+            parallelization of the write-operation and does not work is var_slices = slice(None)
+            (in that casem __getitem__ returns a copy and not a view).
+            This method merges both keys:
+            var[mergeSlices(var, var_slices, data)] = data
+
+            Parameters
+            ----------
+            var : netcdf4.Variable
+                Variable to which data is to be saved.
+            var_slices :
+                Keys to pass to the set-operator.
+            data : dndarray
+                Data to be saved.
+            data_slices: tuple of slices
+                As returned by the data.comm.chunk method.
+
+            Returns
+            -------
+
+                Key for the set-operation.
+
+            Raises
+            -------
+
+            """
+            slices = data_slices
+            if slices is None:
+                _, _, slices = data.comm.chunk(data.gshape, data.split if is_split else 0)
+            start, count, stride, _ = nc.utils._StartCountStride(
+                elem=var_slices,
+                shape=var.shape,
+                dimensions=var.dimensions,
+                grp=var.group(),
+                datashape=data.shape,
+                put=True,
+            )
+            out_shape = nc._netCDF4._out_array_shape(count)
+            out_split = getExpandedSplit(data.shape, out_shape, data.split)
+
+            start, count, stride = start.T, count.T, stride.T  # transpose for iteration
+            stop = start + stride * count
+            new_slices = []
+            for begin, end, step in zip(start, stop, stride):
+                if begin.size == 1:
+                    begin, end, step = begin.item(), end.item(), step.item()
+                    new_slices.append(slice(begin, end, step))
+                else:
+                    begin, end, step = begin.flatten(), end.flatten(), step.flatten()
+                    new_slices.append(
+                        np.r_[
```

Comment:
I did not test any of this on GPUs. I'm using numpy here because I'm working on `start, count, stride` which are generated by `nc.utils` and are ndarrays. That's ok because `var_slices` is usually (a tuple of) slices/ints/lists and the same for every process. You're right though, this will fail when using a dndarray for indexing because a netcdf.Variable does not support to be indexed by a dndarray (results in `IndexError: only integers, slices (`:`), ellipsis (`...`), and 1-d integer or boolean arrays are valid indices`). However, it does support being indexed by a `torch.tensor` so it might be interesting to see what happens when using a gpu-tensor for indexing.

Reply:
Not that its relevant here, but why does indexing with a dndarray behave different than a torch.tensor?
```python
In [1]: import heat as ht                                                                                                                                                                                   
In [2]: import numpy as np                                                                                                                                                                                  
In [3]: import torch                                                                                                                                                                                        
In [4]: arr = ht.arange(3).expand_dims(1) + ht.arange(3)                                                                                                                                                    
In [5]: arr                                                                                                                                                                                                 
Out[5]: 
tensor([[0, 1, 2],
        [1, 2, 3],
        [2, 3, 4]], dtype=torch.int32)

In [6]: ind = [0,1]                                                                                                                                                                                         
In [7]: arr[ind]                                                                                                                                                                                            
Out[7]: 
tensor([[0, 1, 2],
        [1, 2, 3]], dtype=torch.int32)

In [8]: arr[np.array(ind)]                                                                                                                                                                                  
Out[8]: 
tensor([[0, 1, 2],
        [1, 2, 3]], dtype=torch.int32)

In [9]: arr[torch.tensor(ind)]                                                                                                                                                                              
Out[9]: 
tensor([[0, 1, 2],
        [1, 2, 3]], dtype=torch.int32)

In [10]: arr[tuple(ind)]                                                                                                                                                                                    
Out[10]: tensor(1, dtype=torch.int32)

In [11]: arr[ht.array(ind)]                                                                                                                                                                                 
Out[11]: tensor(1, dtype=torch.int32)
```

Reply:
the second is because the advanced indexing using heat DNDarrays is not yet implemented, but that is on the list of things to do.

@Markus-Goetz what is the verdict on NumPy here? im not sure. because its a saving routine it is bound to the CPU anyway. but we would need to specify that higher up in the function. if that is done then i have no problem with it.

Reply:
Have an issue about the advanced indexing. 

Numpy seems fine here, ensure that the data is actually on cpu if necessary